src/Extensions.Hosting.AsyncInitialization/Hosting/AsyncInitializationHostExtensions.cs

```diff
@@ -32,5 +32,35 @@ public static async Task InitAsync(this IHost host, CancellationToken cancellati
 
             await rootInitializer.InitializeAsync(cancellationToken);
         }
+
+
+        /// <summary>
+        /// Initializes and runs the application, by first calling all registered async initializers.
+        /// Upon completion, any registered initializers implementing <see cref="IAsyncTeardown"/> are called.
+        /// </summary>
+        /// <param name="host">The host.</param>
+        /// <param name="cancellationToken">Optionally propagates notifications that the operation should be cancelled</param>
+        /// <returns>A task that represents the initialization completion.</returns>
+        /// <exception cref="ArgumentNullException"></exception>
+        /// <exception cref="InvalidOperationException"></exception>
+        /// <exception cref="OperationCanceledException"></exception>
+        public static async Task RunInitAsync(this IHost host, CancellationToken cancellationToken = default)
+        {
+            if (host == null) throw new ArgumentNullException(nameof(host));
+
+            using var scope = host.Services.CreateScope();
+            var rootInitializer = scope.ServiceProvider.GetService<RootInitializer>()
+                ?? throw new InvalidOperationException("The async initialization service isn't registered, register it by calling AddAsyncInitialization() on the service collection or by adding an async initializer.");
+
+            try
+            {
+                await rootInitializer.InitializeAsync(cancellationToken).ConfigureAwait(false);
+                await host.RunAsync(cancellationToken).ConfigureAwait(false);
+            }
+            finally
+            {
+                await rootInitializer.TeardownAsync().ConfigureAwait(false);
+            }
+        }
```

Comment:
I would also expose a `TeardownAsync` method, in case the user wants to explicitly do the teardown in their Main method.

Reply:
Good idea. Done.

Reply:
Actually, after some more thought and tests, providing this method poses a problem due to the fact that `Host.RunAsync()` will dispose the `IHost ` and the underlying `IServiceProvider `upon termination.

```
public static async Task TeardownAsync(this IHost host, CancellationToken cancellationToken = default)
        {
            if (host == null)
                throw new ArgumentNullException(nameof(host));

            using var scope = host.Services.CreateScope();
            var rootInitializer = scope.ServiceProvider.GetService<RootInitializer>()
                ?? throw new InvalidOperationException("The async initialization service isn't registered, register it by calling AddAsyncInitialization() on the service collection or by adding an async initializer.");

            await rootInitializer.TeardownAsync(cancellationToken).ConfigureAwait(false); 
        }
```

So calling `host.TeardownAsync()` after `host.RunAsync()` in the main method results in an `ObjectDisposedException `on the call to `using var scope = host.Services.CreateScope();`

This problem of course does not arise in RunWithInitAsync() because we retain a reference to the scope/RootInitializer.

One could still use the method in an error handler to teardown after initialization failed, but it only works if host.RunAsync() was not called. I'm not sure if it's worth adding the method considering the constraints, unless you have a better idea :-)

Reply:
Ah, yes, forgot about that.

> This problem of course does not arise in RunWithInitAsync() because we retain a reference to the scope/RootInitializer.

Actually, the problem can still arise. If any of the initializer depends on a disposable service, the service will have been disposed as when when the host is disposed.

Which means, basically, that there's no way to safely execute the teardown after the host has stopped...

Reply:
Actually, it could work, but you would need to manually call `StartAsync` and `WaitForShutdownAsync` instead of calling `RunAsync`. The code for `RunAsync` (at least in 2.1, which this library targets) is pretty simple:

https://github.com/aspnet/Hosting/blob/1da228d28fdfcfb2d77241c4e0413efec1beeafd/src/Microsoft.Extensions.Hosting.Abstractions/HostingAbstractionsHostExtensions.cs#L57

So you could do the teardown before the end of the `using` block.

However, the `RunAsync` code has changed in later versions, as it now handles the case where the host implements `IAsyncDisposable`. And since that interface was introduced in .NET Core 3.0 / .NET Standard 2.1, we can't use it in this library, unless we add another TFM. So, it adds a quite bit of complexity to be able to do this correctly...

(Of course, none of this makes it possible to add a separate `TeardownAsync`, because as you mentioned, once `RunAsync` has returned, it's too late to try to resolve a service)

Reply:
> Actually, the problem can still arise. If any of the initializer depends on a disposable service, the service will have been disposed as well when when the host is disposed.

Well, I did some tests (see latest pull request) and cannot confirm that behavior. To my understanding, objects will only get disposed by the DI container if their reference count is 0. At least in my tests, initializers depending on disposable services still worked after the host is disposed.

Reply:
> Which means, basically, that there's no way to safely execute the teardown after the host has stopped...

I removed the TeardownAsync() extension method as a consequence. 

Actually there would be a way to solve this by means of the `IHostApplicationLifetime.ApplicationStopped` property (which was my very first approach). But that interface is not supported by netstandard and it kinda makes thinks ugly.

Reply:
> Well, I did some tests (see latest pull request) and cannot confirm that behavior.

I just did a test of my own, and I did confirm it. Simple test in Linqpad (just add the Microsoft.Extensions.Hosting package):

```csharp
async Task Main()
{
    var builder = Host.CreateDefaultBuilder();
    builder.ConfigureServices(services =>
    {
       services.AddTransient<MyService>();
       services.AddTransient<MyDisposableService>();
    });
    
    var host = builder.Build();
    host.Services.GetRequiredService<MyService>();
    using var cts = new CancellationTokenSource(TimeSpan.FromSeconds(1));
    Console.WriteLine("Before RunAsync");
    await host.RunAsync(cts.Token);
    Console.WriteLine("After RunAsync");
}


class MyService
{
    public MyService(MyDisposableService disposableService)
    {
    }
}

class MyDisposableService : IDisposable
{
    public void Dispose()
    {
        Console.WriteLine("Disposing MyDisposableService");
    }
}
```

This code prints "Disposing MyDisposableService" before "After RunAsync"

> To my understanding, objects will only get disposed by the DI container if their reference count is 0

There's no reference counting in .NET, but yes, if they're no longer referenced, they will eventually be collected by the GC, which will run the finalizer, if there's one, which might call `Dispose` if it's correctly implemented (that's a lot of ifs). But the DI container doesn't wait for the GC to run; when it's disposed, it also disposes any disposable objects that it created (even if they're still referenced).

> > Which means, basically, that there's no way to safely execute the teardown after the host has stopped...
> 
> I removed the TeardownAsync() extension method as a consequence.

Yes, but because of what I mentioned before, it's still not safe to do it after `RunAsync` has completed, since disposable services would have been disposed. It would work _if_ your teardown doesn't use any disposable service, but not in the general case.

> Actually there would be a way to solve this by means of the `IHostApplicationLifetime.ApplicationStopped` property (which was my very first approach). But that interface is not supported by netstandard and it kinda makes thinks ugly.

It's a bit ugly, yes, but more importantly, it can't be async... The callback you register on the `ApplicationStopped` cancellation token can only be an `Action` or `Action<object?>`, so it wouldn't wait for your async operation to completed.

Reply:
Hi @thomaslevesque ,
You're right that the Dispose() method gets called, but I was still able to access the disposable service in the TeardownAsync() method. 
In any case, it seems as though there's no general purpose implementation that will always work. In my case, teardown is not dependant on on service that would get disposed, so I'm not facing that issue. The way it's implemented now works fine for my usecase. But I understand that you're expecting a general purpose implementation without the disposable restrictions that works in the context of your library. So I guess we need to close the Pull Request...? Unless you have some other idea/approach...
Cheers, Alexander

Reply:
> But I understand that you're expecting a general purpose implementation without the disposable restrictions that works in the context of your library. So I guess we need to close the Pull Request...? Unless you have some other idea/approach...

Well, as I mentioned earlier, it could work by calling `StartAsync` and `WaitForShutdownAsync` yourself, instead of using `RunAsync`. Basically something like this:

```csharp
public static async Task InitAndRunAsync(this IHost host, CancellationToken cancellationToken = default)
{
    using (host)
    {
        // initialization
        ...

        await host.StartAsync(token);
        await host.WaitForShutdownAsync(token);

        // teardown
        ...
    }
}
```

(I assume the services won't actually be disposed until the end of the `using` block, but that should be checked)

What makes things a bit more complex is that in more recent versions, `RunAsync` actually looks [like this](https://github.com/dotnet/runtime/blob/b9db657a0e7789ce97f6cb6683401fd31639d67c/src/libraries/Microsoft.Extensions.Hosting.Abstractions/src/HostingAbstractionsHostExtensions.cs#L63), with a special case for `IAsyncDisposable`. Ideally `InitAndRunAsync` would need to do the same thing. But since `IAsyncDisposable` doesn't exist in .NET Standard 2.0, that would require adding a new TFM to the library.

TBH, I should probably just retarget the library to a newer .NET version. There's no point in continuing to support .NET Core 2.1... I need to think about it.

Reply:
@thomaslevesque 
I did actually already update by fork to .net 5.
What target were you thinking of? I'd be happy to continue on the project. Updateing would surely make sense.
I'll do an implementation as suggested above and will do some further testing.
Let me know what you think.
Cheers, Alexander

Reply:
I personally would go for net7...

Reply:
I was thinking of .NET 6.0, since it's LTS and still supported.

Projects still targeting .NET Core 2.1 or 3.X, or .NET 5.0, are probably not really maintained anymore, so they won't be bothered if they can't upgrade to the latest version of this lib.